aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -0,0 +1,1479 @@
+    /// For latter case, extracted coins will be used to buy shares into the active pool and
+    /// escape inactivation when current lockup ends.
+    fun redeem_inactive_shares(
+        pool: &mut DelegationPool,
+        shareholder: address,
+        coins_amount: u64,
+        lockup_epoch: u64,
+    ): u64 {
+        let current_lockup_epoch = pool.current_lockup_epoch;
+        let inactive_shares = table::borrow_mut(&mut pool.inactive_shares, lockup_epoch);
+
+        let shares_to_redeem = amount_to_shares_to_redeem(inactive_shares, shareholder, coins_amount);
+        if (shares_to_redeem == 0) return 0;
+        let redeemed_coins = pool_u64::redeem_shares(inactive_shares, shareholder, shares_to_redeem);
+
+        // if delegator reactivated entire pending_inactive stake or withdrawn entire past stake,
+        // enable unlocking operation again
+        if (pool_u64::shares(inactive_shares, shareholder) == 0) {
+            table::remove(&mut pool.pending_withdrawal, shareholder);
+        };
+
+        if (lockup_epoch < current_lockup_epoch) {
+            // withdrawing from ended lockup epoch requires total inactive coins to be decreased
+            pool.total_coins_inactive = pool.total_coins_inactive - redeemed_coins;
+
+            // delete shares pool of ended lockup epoch if everyone have withdrawn
+            if (total_coins(inactive_shares) == 0) {
+                pool_u64::destroy_empty(table::remove<u64, pool_u64::Pool>(&mut pool.inactive_shares, lockup_epoch));
+            }
+        };
+
+        redeemed_coins
+    }
+
+    fun synchronize_delegation_pool(pool: &mut DelegationPool) {
+        let pool_address = get_pool_address(pool);
+        let (active, inactive, pending_active, pending_inactive) = stake::get_stake(pool_address);
+
+        // update total coins accumulated by `active` + `pending_active` shares
+        pool_u64::update_total_coins(&mut pool.active_shares, active + pending_active);
+
+        // advance lockup epoch on delegation pool if it already passed on the stake one (AND stake explicitly inactivated)
+        if (last_reconfiguration_time() / MICRO_CONVERSION_FACTOR >= pool.locked_until_secs && pending_inactive == 0) {
+            pool.locked_until_secs = stake::get_lockup_secs(pool_address);
+
+            // `inactive` on stake pool == remaining inactive coins over ended lockup epochs +
+            // `pending_inactive` stake and its rewards (both inactivated) on this ending lockup
+            let ended_lockup_total_coins = inactive - pool.total_coins_inactive;
+            pool_u64::update_total_coins(pending_inactive_shares_pool(pool), ended_lockup_total_coins);
+
+            // capture inactive coins over all ended lockup cycles (including this ending one)
+            pool.total_coins_inactive = inactive;
+
+            // if no `pending_inactive` stake on the ending lockup, reuse its shares pool
```

Comment:
What does reuse mean? Shouldn't the key still be updated to be the current epoch index?

Reply:
well, if the lockup cycle on the stake pool passed by and no stake has been inactivated, we can reuse that shares pool for pending-inactive stake in the new cycle without advancing the 'observed' cycle on delegation pool

Reply:
It seems strange/wrong to not advance it though. Can mess up events/off chain queries if they rely on the observed lockup cycle index